src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Responsible for updating data related to Document outline.
+    /// It is expected that all operations this type do not need to be on the UI thread.
+    /// </summary>
+    internal partial class DocumentOutlineViewModel : INotifyPropertyChanged, IDisposable
+    {
+        private readonly ILanguageServiceBroker2 _languageServiceBroker;
+        private readonly CompilationAvailableTaggerEventSource _textViewEventSource;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly ITextBuffer _textBuffer;
+
+        private CancellationToken CancellationToken => _cancellationTokenSource.Token;
+
+        public DocumentOutlineViewModel(
+            ILanguageServiceBroker2 languageServiceBroker,
+            IAsynchronousOperationListener asyncListener,
+            CompilationAvailableTaggerEventSource textViewEventSource,
```

Comment:
Should this just be the base type of tagger event sources than the explicit base type?